File: spoj/histogra.cpp
#include <functional>
#include <numeric>
#include <iostream>
#include <cstdio>
#include <limits>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <cstring>
#include <cassert>
#include <vector>
#include <list>
#include <stack>
#include <map>
#include <set>
#include <deque>
#include <queue>
#include <bitset>
#include <sstream>
#include <algorithm>
using namespace std;

inline long long int glong()
{
    long long int n = 0;
    int sign=1;
    register char c=0;
    while(c<33)
        c=getchar_unlocked();
    if (c=='-')
    {
        sign=-1;
        c=getchar_unlocked();
    }
    while(c>='0'&&c<='9')
    {
        n=(n<<3)+(n<<1)+(c-'0');
        c=getchar_unlocked();
    }
return n*sign;
}

void plong(long long int a)
{
    register char c;
    char num[20];
    int i = 0;
    if(a < 0){
        putchar_unlocked('-');
        a *= -1;
    }
    do
    {
        num[i++] = a%10 + 48;
        a /= 10;
    }  while (a != 0);
    i--;
    while (i >= 0)
        putchar_unlocked(num[i--]);
    putchar_unlocked('\n');
}
template<class arrtype, class type>
class RMQ
{
    arrtype& arr;
    type* tree;
    size_t size;
private:
    type constructTree(size_t lo, size_t hi, size_t ind)
    {
        if ((hi - lo) == 1)
        {
            tree[ind] = lo;
            return tree[ind];
        }
        size_t mid = lo + ((hi - lo) / 2);
        type tmp1, tmp2;
        tmp1 = constructTree(lo, mid, 2*ind + 1);
        tmp2 = constructTree(mid, hi, 2*ind + 2);
        if (arr[tmp1] <= arr[tmp2])
            tree[ind] = tmp1;
        else
            tree[ind] = tmp2;
        return tree[ind];
    }
    type callQuery(size_t beg, size_t end, size_t qbeg, size_t qend, size_t ind)
    {
        if ((beg >= qbeg) and (end <= qend))
            return tree[ind];

        if (((beg < qbeg) and (end <= qbeg)) or ((beg >= qend) and (end > qend)))
            return -1;

        size_t mid = beg + ((end - beg) / 2);
        type tmp1, tmp2;
        tmp1 = callQuery(beg, mid, qbeg, qend, 2*ind + 1);
        tmp2 = callQuery(mid, end, qbeg, qend, 2*ind + 2);
        if (tmp1 == -1)
            return tmp2;
        else if (tmp2 == -1)
            return tmp1;
        if (arr[tmp1] <= arr[tmp2])
            return tmp1;
        else
            return tmp2;
    }
public:
    RMQ(arrtype& array, size_t array_size) : arr(array)
    {
        size = array_size;
        tree = new type[size_t(2*(pow(2, ceil(log2(size)))) + 1)];
        constructTree(0, size, 0);
    }
    type query(size_t lo, size_t hi)
    {
        return callQuery(0, size, lo, hi, 0);
    }
};

long long int maxarea(vector<long long int>& arr, long long int lo, long long int hi, RMQ<vector<long long int>, long long int>& rangeminq)
{
    if(lo >= hi)
        return 0;
    if (hi - lo == 1)
        return arr[lo];
    if (hi - lo == 2)
        return min(arr[lo], arr[lo + 1])*2;
    long long int ind = rangeminq.query(lo, hi);
    return max(
        max(
            maxarea(arr, lo, ind, rangeminq),
            maxarea(arr, ind + 1, hi, rangeminq)
        ),
        arr[ind] * (hi - lo)
    );
}
int main(int argc, char *argv[])
{
    while(true)
    {
        long long int length = glong();
        long long int maximum_area = 0;
        vector<long long int> lst;
        if (length == 0)
            break;
        for (long long int x = 0; x < length; ++x)
        {
            lst.push_back(glong());
        }
        RMQ<vector<long long int>, long long int> rangeminq(lst, length);
        maximum_area = maxarea(lst, 0, length, rangeminq);
        plong(maximum_area);
    }
    return 0;
}

File: spoj/histogra.py


def main():
    rangeminq = None
    class rmq:

        def __init__(self, arr):
            self.arr = arr
            self.tree = self.gettree()

        def gettree(self):
            arr = self.arr
            from math import log, ceil
            n = len(arr)
            tree = [0] * (2*(pow(2, int(ceil(log(n, 2))))) - 1)

            def _gettree(arr, tree, lo, hi, ind):
                if hi - lo == 1:
                    tree[ind] = arr[lo]
                    return tree[ind]
                mid = lo + (hi - lo) // 2
                tree[ind] = min(
                    _gettree(arr, tree, lo, mid, 2*ind + 1),
                    _gettree(arr, tree, mid, hi, 2*ind + 2))
                return tree[ind]
            _gettree(arr, tree, 0, len(arr), 0)
            return tree

        def updateval(self, ind, val):
            tree = self.tree
            arr = self.arr

            # in this function val and arr are not piggy backed because they
            # are used only once in an update.
            def _updateval(tree, beg, end, i, ind):
                if end - beg == 1:
                    if i == beg:
                        arr[i] = val
                        tree[ind] = val
                        return tree[ind]
                    else:
                        return tree[ind]
                if i < beg or i >= end:
                    return tree[ind]

                mid = beg + (end - beg) // 2
                tree[ind] = min(
                    _updateval(tree, beg, mid, i, 2*ind + 1),
                    _updateval(tree, mid, end, i, 2*ind + 2))
                return tree[ind]
            _updateval(tree, 0, len(arr), ind, 0)

        def query(self, lo, hi):
            def _query(tree, beg, end, qbeg, qend, ind):
                if beg >= qbeg and end <= qend:
                    return tree[ind]

                if (beg < qbeg and end <= qbeg) or (beg >= qend and end > qend):
                    return float('inf')

                mid = beg + (end - beg) // 2
                return min(
                    _query(tree, beg, mid, qbeg, qend, 2*ind + 1),
                    _query(tree, mid, end, qbeg, qend, 2*ind + 2))

            return _query(self.tree, 0, len(self.arr), lo, hi, 0)

    def maxarea(arr, lo, hi):
        if lo >= hi:
            return 0
        if hi - lo < 3:
            return min(arr[lo:hi])*(hi - lo)
        return max(
            maxarea(arr, lo, arr.index(rangeminq.query(lo, hi), lo, hi)),
            maxarea(arr, arr.index(rangeminq.query(lo, hi), lo, hi) + 1, hi),
            rangeminq.query(lo, hi) * (hi - lo)
        )
    while True:
        lst = map(int, raw_input().split())
        length = lst.pop(0)
        if not length:
            break
        if length < 3:
            print min(lst)*length
            continue
        rangeminq = rmq(lst)
        print maxarea(lst, 0, len(lst))

main()
